### PR TITLE
fix(authentication-service): fix refresh token api creating access token with default tenant id.

### LINKED ISSUE #1430: Refresh Token API switches back to default tenant in the token
**Describe the bug**
The refresh Token API switches back to default tenant in the token even when my token has a

**To Reproduce**
1. Add a user in user DB belonging to 2 or more tenants.
2. Create a token for the user having tenantId other than default tenantId
3. hit /token-refresh API
4. New access token will be generated & it will have default tenant id.

**Expected behavior**
 New access token generated should have tenantId same as that of the previous expired access token

### PR DESCRIPTION
**Pull request by Surbhi-sharma1**
fixed refresh token API creating access token with default tenant id.

GH-1430

## Description

- Added auth token switch API that will switch access token's tenant id with the required tenant id.

Fixes #1430 

## Type of change

Please delete options that are not relevant.

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Intermediate change (work in progress)

## How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [ ] Test A
- [ ] Test B

## Checklist:

- [x] Performed a self-review of my own code
- [x] npm test passes on your machine
- [ ] New tests added or existing tests modified to cover all changes
- [ ] Code conforms with the style guide
- [ ] API Documentation in code was updated
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: services/authentication-service/src/modules/auth/login.controller.ts [hunk 1]
```diff
@@ -298,36 +298,15 @@ export class LoginController {
...
-        userId: refreshPayload.userId,
-        externalAuthToken: refreshPayload.externalAuthToken,
-        externalRefreshToken: refreshPayload.externalRefreshToken,
+        clientId: payload.refreshPayload.clientId,
+        userId: payload.refreshPayload.userId,
+        externalAuthToken: payload.refreshPayload.externalAuthToken,
+        externalRefreshToken: payload.refreshPayload.externalRefreshToken,
       },
-      authClient,
+      payload.authClient,
```
Comment (rashisf) [inline services/authentication-service/src/modules/auth/login.controller.ts:309]: we need to pass tenant Id here as well which was the bug as after refreshing token, the new accessToken will again have the default tenant_id.

File: services/authentication-service/src/providers/jwt-payload.provider.ts [hunk 2]
```diff
@@ -44,12 +44,16 @@ export class JwtPayloadProvider implements Provider<JwtPayloadFn> {
   ) {}
 
   value() {
-    return async (authUserData: IAuthUser, authClient: IAuthClient) => {
+    return async (
+      authUserData: IAuthUser,
+      authClient: IAuthClient,
+      tenantId?: string,
```
Comment (rashisf) [inline services/authentication-service/src/providers/jwt-payload.provider.ts:50]: we can make this more generic if you want.
 You can use a free options parameter & there you can pass tenant_id as a property here

**Review by rashisf — CHANGES_REQUESTED** [review]

File: services/authentication-service/src/modules/auth/login.controller.ts (hunk 1 above)
Comment (Surbhi-sharma1) [reply services/authentication-service/src/modules/auth/login.controller.ts:309]: Done

File: services/authentication-service/src/modules/auth/login.controller.ts [hunk 3]
```diff
@@ -441,9 +422,83 @@ export class LoginController {
     return new AuthUser(this.user);
   }
 
+  @authenticate(STRATEGY.BEARER, {
+    passReqToCallback: true,
+  })
+  @authorize({permissions: ['*']})
+  @post('/auth-token-switch', {
```
Comment (samarpan-b) [inline services/authentication-service/src/modules/auth/login.controller.ts:429]: wrong url. It should be /auth/switch-token

File: services/authentication-service/src/modules/auth/login.controller.ts [hunk 4]
```diff
@@ -298,36 +299,16 @@ export class LoginController {
...
-        externalAuthToken: refreshPayload.externalAuthToken,
-        externalRefreshToken: refreshPayload.externalRefreshToken,
+        clientId: payload.refreshPayload.clientId,
+        userId: payload.refreshPayload.userId,
+        externalAuthToken: payload.refreshPayload.externalAuthToken,
+        externalRefreshToken: payload.refreshPayload.externalRefreshToken,
       },
-      authClient,
+      payload.authClient,
+      req,
```
Comment (samarpan-b) [inline services/authentication-service/src/modules/auth/login.controller.ts:311]: only take tenantid here

File: services/authentication-service/src/modules/auth/login.controller.ts [hunk 5]
```diff
@@ -441,9 +422,83 @@ export class LoginController {
...
+    await this.refreshTokenRepo.delete(req.refreshToken);
+    return {
+      refreshPayload: refreshPayload,
+      authClient: authClient,
+    };
+  }
   private async createJWT(
     payload: ClientAuthCode<User, typeof User.prototype.id> & ExternalTokens,
     authClient: AuthClient,
+    options?: JWTOptions,
```
Comment (samarpan-b) [inline services/authentication-service/src/modules/auth/login.controller.ts:501]: Where is this interface ?

**Review by samarpan-b — CHANGES_REQUESTED** [review]

File: services/authentication-service/src/modules/auth/login.controller.ts (hunk 1 above)
Comment (samarpan-b) [reply services/authentication-service/src/modules/auth/login.controller.ts:309]: Its in a separate parameter @rashisf

File: services/authentication-service/src/modules/auth/login.controller.ts (hunk 3 above)
Comment (Surbhi-sharma1) [reply services/authentication-service/src/modules/auth/login.controller.ts]: corrected its url.

File: services/authentication-service/src/providers/jwt-payload.provider.ts (hunk 2 above)
Comment (samarpan-b) [reply services/authentication-service/src/providers/jwt-payload.provider.ts:50]: We want it to be specific intentionally

**Review by samarpan-b — CHANGES_REQUESTED** [review]

File: services/authentication-service/src/modules/auth/login.controller.ts [hunk 6]
```diff
@@ -441,9 +421,83 @@ export class LoginController {
...
+        },
+      },
+      ...ErrorCodes,
+    },
+  })
+  async switchToken(
+    @requestBody() req: AuthRefreshTokenRequest,
+  ): Promise<TokenResponse> {
+    if (!req.tenantId) {
+      throw new HttpErrors.Unauthorized(AuthErrorKeys.InvalidCredentials);
```
[the remark this answers is not in the document]
Comment (Surbhi-sharma1) [reply services/authentication-service/src/modules/auth/login.controller.ts]: Done.

File: services/authentication-service/src/modules/auth/login.controller.ts [hunk 7]
```diff
@@ -298,36 +298,16 @@ export class LoginController {
...
-        externalAuthToken: refreshPayload.externalAuthToken,
-        externalRefreshToken: refreshPayload.externalRefreshToken,
+        clientId: payload.refreshPayload.clientId,
+        userId: payload.refreshPayload.userId,
+        externalAuthToken: payload.refreshPayload.externalAuthToken,
+        externalRefreshToken: payload.refreshPayload.externalRefreshToken,
       },
-      authClient,
+      payload.authClient,
+      req.tenantId,
```
[the remark this answers is not in the document]
Comment (Surbhi-sharma1) [reply services/authentication-service/src/modules/auth/login.controller.ts]: changes Done

**Review by samarpan-b — CHANGES_REQUESTED** [review]

**Pull request merged**